Elide redundant test instruction for signed comparison predicates

ashay opened the pull request:
Prior to this patch, when emitting code for x64, if an instruction
updated the flags register and if the following instruction was an
equals or not-equals comparison with zero, the runtime elided the `test`
instruction for materializing the comparison.

Although correct, this is limiting, since it excludes the predicates
SGT, SGE, SLT, and SLE. This patch extends the optimization to allow the
omission of the `test` instruction for all signed integer comparisons.
We skip unsigned integer comparisons since unsigned comparisons with
zero can be evaluated earlier in the compilation process.

Fix #117866

ashay committed:
Elide redundant test instruction for signed comparison predicates

Prior to this patch, when emitting code for x64, if an instruction
updated the flags register and if the following instruction was an
equals or not-equals comparison with zero, the runtime elided the `test`
instruction for materializing the comparison.

Although correct, this is limiting, since it excludes the predicates
SGT, SGE, SLT, and SLE. This patch extends the optimization to allow the
omission of the `test` instruction for all signed integer comparisons.
We skip unsigned integer comparisons since unsigned comparisons with
zero can be evaluated earlier in the compilation process.

Fix #117866

jakobbotsch commented on src/coreclr/jit/emitxarch.cpp:
```diff
@@ -1464,7 +1464,9 @@ bool emitter::AreFlagsSetToZeroCmp(regNumber reg, emitAttr opSize, GenCondition
     }
 
-    if ((cond.GetCode() == GenCondition::NE) || (cond.GetCode() == GenCondition::EQ))
+    if ((cond.GetCode() == GenCondition::NE) || (cond.GetCode() == GenCondition::EQ) ||
+        (cond.GetCode() == GenCondition::SLE) || (cond.GetCode() == GenCondition::SLT) ||
+        (cond.GetCode() == GenCondition::SGE) || (cond.GetCode() == GenCondition::SGT))
```
The code below requires only the zero flag to be written, but that's not sufficient for these other conditions. I expect that's the cause of the test failures.

ashay replied:
Thanks for the tip!  I've added the checks for other flag bits and separated them by the condition code so that flag checks are specific to the requested condition code.

ashay committed:
Check whether preceding instruction updates required flags

Add new checks for specific flag bits based on the requested condition
predicate

ashay commented:
@dotnet-policy-service agree

xtqqczze commented:
> Perhaps we should check whether the `test` instruction's consumer looks up just OF, SF, ZF, and PF or whether it references more flags than those four.

`adc`, `sbb`, and various jump instructions all reference CF.

ashay commented:
> `adc`, `sbb`, and various jump instructions all reference CF, which `test` sets to zero.

Indeed, you're right, but I was wondering in the context of the SGE, SLE, SGT, and SLT comparisons.  It seems like CF shouldn't matter for those comparisons.

xtqqczze reviewed (commented):
test instruction resets OF
